pkg/rules/adapters/ingest.go

```diff
@@ -0,0 +1,93 @@
+package adapters
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+
+	"github.com/timescale/promscale/pkg/pgmodel/cache"
+	"github.com/timescale/promscale/pkg/pgmodel/ingestor"
+	"github.com/timescale/promscale/pkg/pgmodel/metrics"
+	"github.com/timescale/promscale/pkg/pgmodel/model"
+	"github.com/timescale/promscale/pkg/prompb"
+)
+
+var samplesIngested = metrics.IngestorItems.With(map[string]string{"type": "metric", "kind": "sample", "subsystem": "ruling"})
+
+type ingestAdapter struct {
+	ingestor *ingestor.DBIngestor
+}
+
+func NewIngest(ingestor *ingestor.DBIngestor) *ingestAdapter {
+	return &ingestAdapter{ingestor}
+}
+
+type appenderAdapter struct {
+	// Map allows us to group similar samples into a single prompb.TimeSeries
+	// We can technically just maintain a list and have a single sample in each
+	// []prompb.TimeSeries like how Prometheus writes, but grouping samples gives
+	// us advantage in our ingestor.
+	data     map[string][]model.Insertable
+	ingestor *ingestor.DBIngestor
+}
+
+func (a ingestAdapter) Appender(_ context.Context) storage.Appender {
+	return &appenderAdapter{
+		data:     make(map[string][]model.Insertable),
+		ingestor: a.ingestor,
+	}
+}
+
+func (app *appenderAdapter) Append(_ storage.SeriesRef, l labels.Labels, t int64, v float64) (storage.SeriesRef, error) {
+	series, metricName, err := cache.CreateSeries(labelToPrompbLabels(l))
+	if err != nil {
+		return 0, fmt.Errorf("error creating series: %w", err)
+	}
+
+	samples := model.NewPromSamples(series, []prompb.Sample{{Timestamp: t, Value: v}})
+
+	if _, found := app.data[metricName]; !found {
+		app.data[metricName] = make([]model.Insertable, 0)
+	}
+	app.data[metricName] = append(app.data[metricName], samples)
+	id, _, err := series.GetSeriesID()
+	if err != nil {
+		return 0, nil
+	}
```

Comment:
Sorry, i didnt know it will be error always. The aim here was that series id can be returned the next time, avoiding to create a series. If we have an error doing this optimization, simply don't do it. Which is why returning 0, nil.
But, this optimization is prometheus dependent which right now does not happen. So we can remove this part.